Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,251 @@
+        var remaining: number = files.length - total;
+        console.log('FTP upload failed: ' + message + getFinalStatusMessage() + '\nunprocessed files & directories: ' + remaining);
+    }
+    tl.setResult(tl.TaskResult.Failed, message);
+}
+
+function uploadFiles(): Q.Promise<void> {
+    tl.debug('uploading files');
+
+    var defer: Q.Deferred<void> = Q.defer<void>();
+
+    var outerPromises: Q.Promise<void>[] = []; // these run first, and their then clauses add more promises to innerPromises
+    var innerPromises: Q.Promise<void>[] = [];
+    outerPromises.push(ftpHelper.createRemoteDirectory(remotePath).then(() => {
+        progressDirectoriesProcessed++;
+        printProgress('remote directory successfully created/verified: ' + remotePath);
+    })); // ensure root remote location exists
+
+    files.forEach((file) => {
+        tl.debug('file: ' + file);
+        var remoteFile: string = preservePaths ?
+            path.join(remotePath, file.substring(rootFolder.length)) :
+            path.join(remotePath, path.basename(file));
+
+        remoteFile = remoteFile.replace(/\\/gi, "/"); // use forward slashes always
+        tl.debug('remoteFile: ' + remoteFile);
+
+        var stats = tl.stats(file);
+        if (stats.isDirectory()) { // create directories if necessary
+            outerPromises.push(ftpHelper.createRemoteDirectory(remoteFile).then(() => {
+                progressDirectoriesProcessed++;
+                printProgress('remote directory successfully created/verified: ' + remoteFile);
+            }));
+        } else if (stats.isFile()) { // upload files
+            if (overwrite) {
+                outerPromises.push(ftpHelper.uploadFile(file, remoteFile).then(() => {
+                    progressFilesUploaded++;
+                    printProgress('successfully uploaded: ' + file + ' to: ' + remoteFile);
+                }));
+            } else {
+                outerPromises.push(ftpHelper.remoteExists(remoteFile).then((exists: boolean) => {
+                    if (!exists) {
+                        innerPromises.push(ftpHelper.uploadFile(file, remoteFile).then(() => {
+                            progressFilesUploaded++;
+                            printProgress('successfully uploaded: ' + file + ' to: ' + remoteFile);
+                        }));
+                    } else {
+                        progressFilesSkipped++;
+                        printProgress('skipping file: ' + file + ' remote: ' + remoteFile + ' because it already exists');
+                    }
+                }));
+            }
+        }
+    });
+
+    Q.all(outerPromises).then(() => {
+        Q.all(innerPromises).then(() => {
+            defer.resolve(null);
+        }).fail((err) => {
+            defer.reject(err);
+        })
+    }).fail((err) => {
+        defer.reject(err);
+    });
+
+    return defer.promise;
+}
+
+function doWork() {
```

Comment:
Much better.  Let's have doWork be the only method in the mail file so folks can get the gist of what the task is doing.  

That means move the function above this to another file/module.